### DISCUSSION TITLE
At what point does accessing the request become unsafe?

### DISCUSSION THREAD

**Discussion by kyle-mccarthy** [uNetworking/uWebSockets.js#84] [Q&A] [answered]
I am running into the error `Using uWS.HttpRequest past its request handler return is forbidden (it is stack allocated)`, and am having some trouble figuring out when access to the request is disallowed.  

Right now, I am actually getting the error while still in the request handler. I am not completely sure, but it seems like I lose access to the request once I call `res.onData`, but that may just be a coincidence.

I modified the JsonPost example to use a promise rather than the callback pattern and was able to reproduce it -- here is gist that demonstrates the behavior that I am experiencing
https://gist.github.com/kyle-mccarthy/505904449c3b154e400a8c8a1cc26ff1

Also I am using v15.5 currently and node 10.14.1.

**Comment by ghost** [comment]
Await is technically a return. The async function is technically split in two actual functions; the first part up until await and the continuation is technically its own separate function.

**Comment by ghost** [comment, marked as answer]
Easiest rule of thumb is to just read all headers you need, at the top of the function:

let userAgent = req.getHeader('user-agent');

Then forget about the req and only use userAgent.
